content/oci/storage.go

```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("%s: %s: %w", target.Digest, target.MediaType, errdef.ErrNotFound)
+		}
+		return nil, err
+	}
+
+	return fp, nil
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (s *Storage) Push(_ context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	// calculate the target path for the content.
+	target, err := s.blobPath(expected.Digest)
+	if err != nil {
+		return fmt.Errorf("%s: %s: %w", expected.Digest, expected.MediaType, errdef.ErrInvalidDigest)
+	}
+
+	// check if the target content exists in the blob directory.
+	if _, err := os.Stat(target); err == nil {
+		return fmt.Errorf("%s: %s: %w", expected.Digest, expected.MediaType, errdef.ErrAlreadyExists)
+	}
+
+	if err := ensureDir(filepath.Dir(target)); err != nil {
+		return err
+	}
+
+	// write the content to a temporary ingest file.
+	ingest, err := s.ingest(expected, content)
+	defer cleanUpFile(ingest)
+	if err != nil {
+		return err
+	}
+
+	// move the content from the temporary ingest file to the target path.
+	// the target will be replaced if it already exists.
+	return os.Rename(ingest, target)
+}
+
+// Exists returns true if the described content Exists.
+func (s *Storage) Exists(_ context.Context, target ocispec.Descriptor) (bool, error) {
+	path, err := s.blobPath(target.Digest)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+
+	return false, err
+}
+
+// ingest write the content into a temporary ingest file.
+func (s *Storage) ingest(expected ocispec.Descriptor, content io.Reader) (string, error) {
+	if err := ensureDir(s.ingestRoot); err != nil {
+		return "", err
+	}
+
+	// create a temp file with the file name format "blobDigest_randomString"
+	// in the ingest direcotry.
+	// Go ensures that multiple programs or goroutines calling CreateTemp
+	// simultaneously will not choose the same file.
+	fp, err := os.CreateTemp(s.ingestRoot, expected.Digest.Encoded()+"_*")
+	if err != nil {
+		return "", fmt.Errorf("failed to create ingest file: %v", err)
+	}
+	defer fp.Close()
+
+	path := fp.Name()
+	if err := ioutil.Copy(fp, content, expected); err != nil {
+		return path, fmt.Errorf("failed to ingest: %v", err)
+	}
+
+	// change to readonly
+	if err := os.Chmod(path, 0400); err != nil {
+		return path, fmt.Errorf("failed to make readonly: %v", err)
+	}
+
+	return path, nil
+}
+
+// blobPath calculates blob path from the given digest.
+func (s *Storage) blobPath(dgst digest.Digest) (string, error) {
+	if err := dgst.Validate(); err != nil {
+		return "", fmt.Errorf("cannot calculate blob path from invalid digest %s: %v", dgst.String(), err)
+	}
+
+	return filepath.Join(s.blobRoot, dgst.Algorithm().String(), dgst.Encoded()), nil
+}
+
+// ensureDir ensures the directories of the path exists.
+func ensureDir(path string) error {
+	return os.MkdirAll(path, 0777)

```

Comment:
Does it require read/write/execute for everyone?

Reply:
No, but this will be mased by [umask](https://man7.org/linux/man-pages/man2/umask.2.html)